### DISCUSSION TITLE
Supported Ruby Versions, what Ruby version are you using?

### DISCUSSION THREAD

**Discussion by MSP-Greg** [puma/puma#3390] [Polls]
Given that Ruby 3.0 is EOL, wondering about what Ruby version people are using with Puma 6.x.  Puma currently supports Ruby 2.4 and later.

**Comment by dentarg** [comment]
Seems like Rails 6.1 is still supported (soon EoL?) and then it is possible to still be on Ruby 2.5: https://github.com/minitest/minitest/blob/7ad0ff0d91b1423222770af70919ce2d4d25a03f/README.rdoc#what-versions-are-compatible-with-what-or-what-versions-are-supported-

**Comment by MSP-Greg** [comment]
On 14-Oct-2022, Puma 6.0.0 was released and the minimum Ruby version was changed from 2.2 to 2.4.

Is it time to update Puma’s minimum Ruby version?

Nio4r master is now supporting Ruby 2.6 and later.

One choice would be to set the minimum to Ruby 2.6.  Ruby 2.6.0 was released in Dec-2018.  We might consider pushing the minimum further.  Some Ruby history info below.

Currently, Ruby 3.1 is in “security maintenance phase”, and Ruby 3.2 is in standard support (bug & security fixes).

```text
First Release            Most Recent Release
2022-12-25  Ruby 3.3.0   2024-09-03  3.3.5
2022-12-25  Ruby 3.2.0   2024-07-26  3.2.5
2021-12-25  Ruby 3.1.0   2024-05-29  3.1.6
2020-12-25  Ruby 3.0.0   2024-04-23  3.0.7
2019-12-25  Ruby 2.7.0   2023-03-30  2.7.8
2018-12-25  Ruby 2.6.0   2022-04-12  2.6.10
2017-12-25  Ruby 2.5.0   2021-04-05  2.5.9
2016-12-25  Ruby 2.4.0   2020-03-31  2.4.10
```

**Reply by dentarg** [reply]
Let's set it to 3.1?

**Comment by MSP-Greg** [comment]
> Let's set it to 3.1?

That would be nice.  Wondering if 2.7 should be considered, as it matches Rails 7.0?

It's probably too much, but one feature that would be nice is the 3.2 update `Queue#pop(non_block = false, timeout: nil)`.  The timeout addition may make locking in the ThreadPool a bit simpler...

**Reply by eregon** [reply]
> make locking in the ThreadPool a bit simpler...

That may potentially be helped by https://github.com/ruby-concurrency/concurrent-ruby/pull/1082 once it's merged & released, but I haven't looked at the usage in Puma.

**Comment by dentarg** [comment]
@nateberkopec posted a reply in January that I somehow missed https://github.com/puma/puma/issues/3588#issuecomment-2617138399 tl;dr Puma is very conservative
